### ISSUE TITLE
Not working for VS Code workspaces?

### ISSUE THREAD

**Issue by LaurensBosscher** [charliermarsh/ruff-vscode#59]
First of all, thanks for writing this extension. I think it holds great promise for the future. I've tried to run this on our workspace, consisting of several Python libraries and a Django project but when saving a file I get the pop-up which never completes and only saves the file when I click on "Cancel":

![image](https://user-images.githubusercontent.com/13363196/208083623-0abdfa59-8397-4789-bb3f-1b23fc73693f.png)

I see the following output for the "Ruff" extension:

![image](https://user-images.githubusercontent.com/13363196/208083857-f3940e45-f3a0-45c5-b516-025ce843cb5c.png)

Executing Ruff from the command-line seems to work:

![image](https://user-images.githubusercontent.com/13363196/208084048-87883b10-e5a8-4277-807f-3cc752291806.png)

Executing the bundled Ruff also seems to work:
![image](https://user-images.githubusercontent.com/13363196/208084619-0f0595ac-b3aa-40de-9b26-856647f8e918.png)

I have ruff configured as following:
![image](https://user-images.githubusercontent.com/13363196/208084687-5aa44cbb-ae43-4d46-a1da-7e1eef9f4276.png)

If I open the folder directly in VS Code (so not as part of the workspace) I do see that Ruff is working:
![image](https://user-images.githubusercontent.com/13363196/208095114-aee59a39-2778-4e9e-995e-112e68fd0046.png)

There's no additional output in the VS Code output panel:
![image](https://user-images.githubusercontent.com/13363196/208095249-ef1a6154-0716-4adf-b1fd-542a5816f519.png)

**Comment by charliermarsh** [comment]
Thanks for the clear write-up! If you uninstall Ruff, but keep isort installed, this behavior goes away, right? (Only asking because the pop-up mentions isort too.)

**Comment by charliermarsh** [comment]
I'm going to push some improvements to the workspace settings stuff, would love for you to get it in a try with the next release and report back!

**Comment by LaurensBosscher** [comment]
@charliermarsh, looking forward to the next release! If I can further contribute by doing any debugging please do not hesitate to reach out to me.

I've just tried removing ruff and keeping isosort and that takes care of the issue. As far as I've been to able to isolate it only occurs when I use ruff in combination with a VS Code workspace.

**Comment by charliermarsh** [comment]
@LaurensBosscher - Do you mind trying out `2022.0.23`? If you're still seeing issues, I'll spend some time trying to repro myself!

**Comment by LaurensBosscher** [comment]
@charliermarsh, I've run into the same issue with the new version. 

I've had some time to look into this. I've replaced the "--quit" flag with "--verbose". Perhaps this could be done automatically if the user selects the "debug" logging level to aid in debugging? I'm now getting the following output:

```
Name: Ruff
Module: ruff
Python extension loading
Waiting for interpreter from python extension.
Python extension loaded
Server Command [RUN]: C:\Users\lbo\AppData\Local\Programs\Python\Python310\python.exe c:\Users\lbo\.vscode\extensions\charliermarsh.ruff-2022.0.23-win32-x64\bundled\tool\server.py
Server Command [DEBUG]: C:\Users\lbo\AppData\Local\Programs\Python\Python310\python.exe c:\Users\lbo\.vscode\extensions\charliermarsh.ruff-2022.0.23-win32-x64\bundled\tool\_debug_server.py
Cancel notification for unknown message id "1"
Cancel notification for unknown message id "2"
Failed to handle request 3 textDocument/codeAction CodeActionParams(text_document=TextDocumentIdentifier(uri='file:///c%3A/develop/savex/projects/investment_decision_support_system/bloomberg/views.py'), range=0:0-51:0, context=CodeActionContext(diagnostics=[Diagnostic(range=4:5-4:39, message='Import "BloombergPortfolioModels.portfolio" could not be resolved', severity=<DiagnosticSeverity.Warning: 2>, code='reportMissingImports', code_description=CodeDescription(href='https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportMissingImports'), source='Pylance', tags=None, related_information=None, data=None), Diagnostic(range=2:17-2:19, message='"pd" is not accessed', severity=<DiagnosticSeverity.Hint: 4>, code=None, code_description=None, source='Pylance', tags=[<DiagnosticTag.Unnecessary: 1>], related_information=None, data=None)], only=['source.organizeImports'], trigger_kind=None), work_done_token=None, partial_result_token=None)
Traceback (most recent call last):
  File "c:\Users\lbo\.vscode\extensions\charliermarsh.ruff-2022.0.23-win32-x64\bundled\libs\pygls\protocol.py", line 337, in _handle_request
    self._execute_request(msg_id, handler, params)
  File "c:\Users\lbo\.vscode\extensions\charliermarsh.ruff-2022.0.23-win32-x64\bundled\libs\pygls\protocol.py", line 261, in _execute_request
    self._send_response(msg_id, handler(params))
  File "c:\Users\lbo\.vscode\extensions\charliermarsh.ruff-2022.0.23-win32-x64\bundled\tool\server.py", line 253, in code_action
    results = _formatting_helper(text_document, select="I001")
  File "c:\Users\lbo\.vscode\extensions\charliermarsh.ruff-2022.0.23-win32-x64\bundled\tool\server.py", line 345, in _formatting_helper
    result = _run_tool_on_document(
  File "c:\Users\lbo\.vscode\extensions\charliermarsh.ruff-2022.0.23-win32-x64\bundled\tool\server.py", line 598, in _run_tool_on_document
    rpc_result = jsonrpc.run_over_json_rpc(
  File "c:\Users\lbo\.vscode\extensions\charliermarsh.ruff-2022.0.23-win32-x64\bundled\tool\jsonrpc.py", line 249, in run_over_json_rpc
    return RpcRunResult(data["result"], "", data["error"])
KeyError: 'result'
Cancel notification for unknown message id "4"
Cancel notification for unknown message id "3"
Cancel notification for unknown message id "5"
[Error - 09:40:30] Request textDocument/codeAction failed.
  Message: KeyError: 'result'
  Code: -32602 
{'traceback': ['  File "c:\\Users\\lbo\\.vscode\\extensions\\charliermarsh.ruff-2022.0.23-win32-x64\\bundled\\libs\\pygls\\protocol.py", line 337, in _handle_request\n    self._execute_request(msg_id, handler, params)\n', '  File "c:\\Users\\lbo\\.vscode\\extensions\\charliermarsh.ruff-2022.0.23-win32-x64\\bundled\\libs\\pygls\\protocol.py", line 261, in _execute_request\n    self._send_response(msg_id, handler(params))\n', '  File "c:\\Users\\lbo\\.vscode\\extensions\\charliermarsh.ruff-2022.0.23-win32-x64\\bundled\\tool\\server.py", line 253, in code_action\n    results = _formatting_helper(text_document, select="I001")\n', '  File "c:\\Users\\lbo\\.vscode\\extensions\\charliermarsh.ruff-2022.0.23-win32-x64\\bundled\\tool\\server.py", line 345, in _formatting_helper\n    result = _run_tool_on_document(\n', '  File "c:\\Users\\lbo\\.vscode\\extensions\\charliermarsh.ruff-2022.0.23-win32-x64\\bundled\\tool\\server.py", line 598, in _run_tool_on_document\n    rpc_result = jsonrpc.run_over_json_rpc(\n', '  File "c:\\Users\\lbo\\.vscode\\extensions\\charliermarsh.ruff-2022.0.23-win32-x64\\bundled\\tool\\jsonrpc.py", line 249, in run_over_json_rpc\n    return RpcRunResult(data["result"], "", data["error"])\n']}
Ignoring notification for unknown method "textDocument/didSave"
```

I wasn't able to print the "data" so I changed jsonrpc.py to be a bit more defensive (happy to create a PR!):

![image](https://user-images.githubusercontent.com/13363196/208626001-76bfb783-7f91-418a-81a0-48c3230ebf67.png)

And I got the following error:

![image](https://user-images.githubusercontent.com/13363196/208626101-92589ce7-b136-4b82-bdb1-0f97e284ee7d.png)

I've installed Ruff globally (to test) and I could import Ruff:
![image](https://user-images.githubusercontent.com/13363196/208626595-f2f11ece-64a5-4d4b-86f5-25c4d4e896c9.png)

At this point I got stuck, the only thing that I could think of is that perhaps the "useBundled" strategy changes the path that Python uses for installed modules? 

Is there any way I could make the output of print visible and/or use any other kind of logging?

**Comment by LaurensBosscher** [comment]
I'm not sure if this helps or is relevant at all, but the VS Code Mypy extension includes a function that gets the active interpreter, I've tested it on our Workspace and it seems to be a robust implementation. I'm not nearly as knowledgeable about extensions as I would like to be but perhaps it's useful to take a look at their approach? 

Link to the function is here:
https://github.com/matangover/mypy-vscode/blob/611e5dfaafdce58a5ac84c0aa5d4216ab225487e/src/extension.ts#L665

**Comment by charliermarsh** [comment]
Awesome, thank you for all of this! I’ll dig in later today and figure out what’s going on + how to resolve.

**Comment by charliermarsh** [comment]
(Looking into this now.)

**Comment by charliermarsh** [comment]
As you pointed out above, this does seem to be related to the JSON RPC and `runpy` stuff. Not sure what's going wrong though yet, it's really tedious to debug given all the indirection.

**Comment by charliermarsh** [comment]
I may just get rid of the JSON RPC stuff entirely. We only need the path to the executable.

**Comment by charliermarsh** [comment]
Ok, I think this should be fixed by #67, but let's keep an eye on it. Do you mind testing after I cut the next release (probably later today)?

**Comment by LaurensBosscher** [comment]
> Ok, I think this should be fixed by #67, but let's keep an eye on it. Do you mind testing after I cut the next release (probably later today)?

Happy to do so! I'll keep an eye on the updates and report back after I had a chance to test this.

**Comment by bigluck** [comment]
Ciao @charliermarsh, I've just tested the fix locally, and it works! Well done!

<img width="729" alt="Screenshot 2022-12-21 at 11 53 58" src="https://user-images.githubusercontent.com/1511095/208888665-983ab052-0619-4d24-8cde-b7be6d251da0.png">

**Comment by charliermarsh** [comment]
Amazing, thanks @bigluck!

**Comment by charliermarsh** [comment]
@LaurensBosscher - The change that I think _should_ fix this is live in 2022.0.24.

**Comment by LaurensBosscher** [comment]
@charliermarsh.

Many thanks for the quick fix! It works:
![image](https://user-images.githubusercontent.com/13363196/209792904-0ed69707-61f0-4788-9931-68cace46d7b7.png)
